tmt/steps/prepare/artifact/__init__.py

```diff
@@ -52,16 +58,84 @@ def go(
         logger: Logger,
     ) -> PluginOutcome:
         outcome = super().go(guest=guest, environment=environment, logger=logger)
-        # TODO: Get and handle repositories
-        # TODO: Create the local repository
+
+        # Prepare a shared directory on the guest for aggregating artifacts.
+        shared_repo_dir = self.plan_workdir / 'artifact-shared-repo'
+
+        # Ensure the shared repository directory exists on the guest.
+        shared_repository = create_repository(
+            artifact_dir=shared_repo_dir,
+            guest=guest,
+            logger=logger,
+            repo_name="tmt-artifact-shared",
+        )
+
+        # Initialize all providers and have them contribute to the shared repo
+        providers: list[ArtifactProvider[ArtifactInfo]] = []
         for raw_provider_id in self.data.provide:
-            provider_class = get_artifact_provider(raw_provider_id)
-            provider_id_sanitized = tmt.utils.sanitize_name(raw_provider_id, allow_slash=False)
-            logger = self._logger.descend(raw_provider_id)
-            provider = provider_class(raw_provider_id, logger=logger)
-            download_path = self.plan_workdir / "artifacts" / provider_id_sanitized
-            # TODO: Not using exclude_pattern yet.
-            provider.fetch_contents(guest, download_path)
+            try:
+                provider_class = get_artifact_provider(raw_provider_id)
+
+                # Sanitize the provider ID to use as a directory name
+                provider_id_sanitized = tmt.utils.sanitize_name(raw_provider_id, allow_slash=False)
+                provider_logger = self._logger.descend(raw_provider_id)
+                provider = provider_class(raw_provider_id, logger=provider_logger)
+                providers.append(provider)
+
+                # Define a unique download path for this provider's artifacts
+                download_path = self.plan_workdir / "artifacts" / provider_id_sanitized
+
+                # For providers that manage repositories, skip download/contribution here.
+                # They will discover packages after repository installation.
+                if not provider.get_repositories():
+                    # First, fetch the contents (download artifacts)
+                    provider.fetch_contents(guest, download_path)
```

Comment:
This will break the logic for `KojiProvider` because it needs to do these steps and it has an empty `get_repositories` (as it should because it does not provide a different repository). The logic under here should be completely independent of `get_repositories` and always run.